src/documentation/sdks/euphoria.md

```diff
@@ -18,72 +18,530 @@ See the License for the specific language governing permissions and
-    .using(p -> p.getFirst() + ": " + p.getSecond())
+    .using(p -> p.getKey() + ": " + p.getValue())
     .output();
 
 // Transform Dataset back to PCollection. It can be done in any step of this flow.
 PCollection<String> outputCollection = flow.unwrapped(output);
 
 // Now we can again use Beam transformation. In this case we save words and their count
 // into the text file.
-outputCollection.apply(TextIO.write().to(options.getOutput()));
+outputCollection.apply(TextIO.write().to("counted_words"));
 
 pipeline.run();
 ```
 
+## Euphoria Guide
+### `Flow` your data
+Euphoria API is composed as a flow of operators called `BeamFlow`. There is a set of operators which allows to construct `BeamFlow` according to your Big Data application needs. `BeamFlows` are easily created from Beam `Pipeline`.
+```java
+Pipeline pipeline = Pipeline.create(options);
+
+BeamFlow flow = BeamFlow.of(pipeline);
+```
+Or from `PCollection`. That is especially useful when some of many Beam's IO are utilized as data inputs.
+```java
+PCollection<T> inputPCollection = ...
+
+BeamFlow flow = BeamFlow.of(inputPCollection);
+```
+### Datasets
+Euphoria use concept of 'Datasets' to describe flows of data between operators. This concept is similar to Beam's `PCollection` and can be converted both ways through `BeamFlow`.
+```java
+PCollection<T> someCollection = ...
+
+// PCollection -> Dataset
+Dataset<T> dataset = flow.wrapped(someCollection);
+
+//And now back: Dataset -> PCollection
+PCollection<T> collection = flow.unwrapped(dataset);
+```
+One should not mix different instances of `BeamFlow` when wrapping/unwrapping `PCollections`.
+
+### Inputs and Outputs
+Input data can supplied through Beams IO into `PCollection`, the same way as in Beam, and latter wrapped by `BeamFlow` into `Dataset`.
+
+```java
+PCollection<String> input =
+  pipeline
+    .apply(Create.of("mouse", "rat", "elephant", "cat", "X", "duck"))
+    .setTypeDescriptor(TypeDescriptor.of(String.class));
+
+Dataset<String> dataset = flow.wrapped(input);
+```
+Output can be treated the same way as inputs, last `Dataset` is converted to `Pcollection` and dumped to appropriate IO.
+
+### Adding Operators
+Real power of Euphoria API is in its [operators suite](#operators-reference). Once we get our hand on `Dataset` we are able to create and connect operators. Each Operator consumes one or more input and produces one output
+`Dataset`. Lets look at simple `MapElements` example.
+
+```java
+Dataset<Integer> input = ...
+
+Dataset<String> mappedElements =
+  MapElements
+    .named("Int2Str")
+    .of(input)
+    .using(String::valueOf)
+    .output();
+```
+The operator consumes `input`, it applies given lambda expression (`String::valueOf`) on each element of `input` and returns mapped `Datatset`. Developer is guided through series of steps when creating operator so the declaration of an operator is straightforward. To start building operator just wrote its name and '.'. Your IDE will give you hits.
```

Comment:
Typo: Datatset

```
To start building operator just write its name and your favorite IDE will give you hints what to do.

Reply:
Yeah, then you hit it back and wish for less violent IDE. :D
Fixed.